pkg/reconciler/resolver/taskrun_resolver_reconciler.go

```diff
@@ -0,0 +1,133 @@
+/*
+Copyright 2021 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resolver
+
+import (
+	"context"
+
+	"github.com/tektoncd/pipeline/internal/resolution"
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	clientset "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	listers "github.com/tektoncd/pipeline/pkg/client/listers/pipeline/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+)
+
+type TaskRunResolverReconciler struct {
+	// Implements reconciler.LeaderAware
+	reconciler.LeaderAwareFuncs
+
+	kubeClientSet     kubernetes.Interface
+	pipelineClientSet clientset.Interface
+	taskrunLister     listers.TaskRunLister
+	configStore       reconciler.ConfigStore
+}
+
+const defaultTaskRunResolutionErrorReason = resolution.ReasonTaskRunResolutionFailed
+
+var _ controller.Reconciler = &TaskRunResolverReconciler{}
+
+func (r *TaskRunResolverReconciler) Reconcile(ctx context.Context, key string) error {
```

Comment:
maybe this is happening somewhere else but im wondering if these new reconcilers should check if `experimentalDisableRefResolution` is set before they do anything?

im definitely over thinking it but im wondering what will happen if the controller is configured to start these but `experimentalDisableRefResolution` is false - i.e. both the existing and the new reconcilers try to resolve the refs at the same time?

Reply:
This is a great point.  They'd collide spectacularly. I think for this reason it would be better to have two CLI flags and eliminate the ConfigMap entry.  That way we can error out early during controller startup if the user incorrectly sets `-experimental-enable-resolution-reconcilers` _without_ switching off the existing resolution machinery.